Platform/ResetNvramEntry/ResetNvramEntry.c

```diff
@@ -0,0 +1,168 @@
+/** @file
+  Boot entry protocol implementation of Reset NVRAM boot picker entry.
+
+  Copyright (c) 2022, Mike Beaton. All rights reserved.<BR>
+  SPDX-License-Identifier: BSD-3-Clause
+**/
+
+#include <Guid/AppleVariable.h>
+
+#include <Uefi.h>
+#include <Library/BaseLib.h>
+#include <Library/OcDeviceMiscLib.h>
+#include <Library/UefiBootServicesTableLib.h>
+#include <Library/UefiRuntimeServicesTableLib.h>
+
+#include <Protocol/OcBootEntry.h>
+
+#define OC_MENU_RESET_NVRAM_ENTRY  "Reset NVRAM"
+
+STATIC BOOLEAN  mIsNative     = FALSE;
+STATIC BOOLEAN  mPreserveBoot = FALSE;
+
+STATIC OC_PICKER_ENTRY  mResetNvramBootEntries[1] = {
+  {
+    .Id            = "reset_nvram",
+    .Name          = OC_MENU_RESET_NVRAM_ENTRY,
+    .Path          = NULL,
+    .Arguments     = NULL,
+    .Flavour       = OC_FLAVOUR_RESET_NVRAM,
+    .Auxiliary     = TRUE,
+    .Tool          = FALSE,
+    .TextMode      = FALSE,
+    .RealPath      = FALSE,
+    .SystemAction  = NULL, ///< overridden in set up
+    .ActionConfig  = &mPreserveBoot,
+    .AudioBasePath = OC_VOICE_OVER_AUDIO_FILE_RESET_NVRAM,
+    .AudioBaseType = OC_VOICE_OVER_AUDIO_BASE_TYPE_OPEN_CORE
+  }
+};
+
+STATIC
+EFI_STATUS
+EFIAPI
+ResetNvramGetBootEntries (
+  IN           OC_PICKER_CONTEXT  *PickerContext,
+  IN     CONST EFI_HANDLE         Device,
+  OUT       OC_PICKER_ENTRY       **Entries,
+  OUT       UINTN                 *NumEntries
+  )
+{
+  //
+  // Custom entries only.
+  //
+  if (Device != NULL) {
+    return EFI_NOT_FOUND;
+  }
+
+  DEBUG ((DEBUG_INFO, "BEP: Adding Reset NVRAM entry, preserve boot %u, native %u\n", mPreserveBoot, mIsNative));
+
+  *Entries    = mResetNvramBootEntries;
+  *NumEntries = ARRAY_SIZE (mResetNvramBootEntries);
+
+  return EFI_SUCCESS;
+}
+
+STATIC
+OC_BOOT_ENTRY_PROTOCOL
+  mResetNvramBootEntryProtocol = {
+  OC_BOOT_ENTRY_PROTOCOL_REVISION,
+  ResetNvramGetBootEntries,
+  NULL
+};
+
+//
+// Use OpenCore NVRAM reset, potentially preserving Boot#### entries.
+// Remarks: We don't need to use LoadOptions here, since it is referring to
+// a variable we have access to, but we want to confirm that the system works.
+//
+STATIC
+EFI_STATUS
+InternalSystemActionResetNvram (
+  IN     VOID  *Config
+  )
+{
+  ASSERT (Config == &mPreserveBoot);
+  return OcResetNvram (*((BOOLEAN *)Config));
```

Comment:
Considering the ASSERT, maybe just use `mPreserveBoot`?

Reply:
Yeees. The ASSERT is just to check that the approach is working, and the deref meant to indicate how something would typically use this (if there were multiple callbacks, presumably the best a method like this could do would be to ASSERT if NULL, I guess).